### PR TITLE
Add docs.rs metadata to display features document prove APIs

### PR DESCRIPTION
**Pull request by nategraf**
This PR added Cargo metadata and crate-level attributes to get docs building with additional information on docs.rs. In particular, this will get feature-gates APIs (e.g. the `Prover` trait) to show up on docs.rs.

It also introduces a new CI job that builds the docs using the docs.rs `--cfg` flag, and on a runner _without_ installing `cargo risczero` or other custom tool. This job is intended to check that our docs will build on the docs.rs machines, and we don't get errors about broken docs build (e.g https://docs.rs/crate/cargo-risczero/0.21.0)

### CODE REVIEW

**Comment by flaub** [comment]
Related: https://github.com/risc0/risc0/issues/919

**Comment by nategraf** [comment]
We've now merged docs.rs feature tags to the `risc0-ethereum` repo. It works for the subset of crates that we do publish to `crates.io` right now, with the blocker on the rest being a way to handle the fact that building the contracts crate requires Forge to be installed.  https://github.com/risc0/risc0-ethereum/pull/78

**Comment by SchmErik** [comment]
Hey @nategraf I made a little more progress here: feel free to take any part of it. the `doc_auto_cfg` thing requires you to split out the `#[cfg(feature = ...)]` into a separate line so that's what's going on in this PR to expose the `export-syscalls` feature

**Comment by nategraf** [comment]
> Hey @nategraf I made a little more progress here: feel free to take any part of it. the `doc_auto_cfg` thing requires you to split out the `#[cfg(feature = ...)]` into a separate line so that's what's going on in this PR to expose the `export-syscalls` feature

I'm actually not sure what the issue is here that you are referring to. I'm guessing something is not showing up that should?

File: risc0/zkp/src/prove/prover.rs [hunk 1]
```diff
@@ -40,7 +40,6 @@ fn make_coeffs<H: Hal>(hal: &H, witness: &H::Buffer<H::Elem>, count: usize) -> H
     // Do interpolate
     hal.batch_interpolate_ntt(&coeffs, count);
     // Convert f(x) -> f(3x), which effective multiplies coefficients c_i by 3^i.
-    #[cfg(not(feature = "circuit_debug"))]
```
Comment (flaub) [inline risc0/zkp/src/prove/prover.rs:43]: This is actually useful for proof system development. In fact it'd be good to go further and make this feature flag work on other areas where we introduce entropy.

File: risc0/binfmt/src/image.rs [hunk 2]
```diff
@@ -52,17 +52,30 @@ struct PersistentPageTableInfo {
...
     pub page_table_addr: u32,
     _page_table_size: u32,
+    /// Address of the root page, which is the top layer of the Merkle tree.
     pub root_addr: u32,
+    /// Page index of the root page.
     pub root_idx: u32,
     root_page_addr: u32,
+    /// Total number of pages covered by this page table.
     pub num_pages: u32,
+    /// Number of entires in the root page. The root page may not be full, if the memory space is
```
Comment (flaub) [inline risc0/binfmt/src/image.rs:77]:
```suggestion
    /// Number of entries in the root page. The root page may not be full, if the memory space is
```

File: risc0/binfmt/src/lib.rs [hunk 3]
```diff
@@ -15,6 +15,9 @@
 //! Manages formatted binaries used by the RISC Zero zkVM
 
 #![cfg_attr(all(not(feature = "std"), not(test)), no_std)]
+#![deny(missing_docs)]
+#![deny(rustdoc::broken_intra_doc_links)]
```
Comment (flaub) [inline risc0/binfmt/src/lib.rs:19]: Thank you!

File: risc0/zkp/src/prove/prover.rs (hunk 1 above)
Comment (nategraf) [reply risc0/zkp/src/prove/prover.rs:43]: Ok, I'll add it back and add it to the feature flags listed in `Cargo.toml`

Comment (nategraf) [reply risc0/zkp/src/prove/prover.rs]: Added it back in ffb6c219fc121a7307687c159b419dd98da229e1

**Comment by nategraf** [comment]
@SchmErik and I discussed this and figured out where the confusion came from. This is ready for review / ready to merge when it looks good to you @flaub

**Review by flaub — APPROVED** [review]

**Comment by flaub** [comment]
Are we still running into any issues with this? I see continued changes

**Comment by nategraf** [comment]
> Are we still running into any issues with this? I see continued changes

https://github.com/risc0/risc0/pull/1880 got merged into this PR, and this PR started to fail because CI was running into two issues that stem from that PR. 1) I needed to run `cargo rdme` 2) I needed to update the test workflow since I removed the `docker` feature flag from `risc0-groth16`.

**Pull request merged**